nep-12.mediawiki

```diff
@@ -0,0 +1,2130 @@
+  switch(type) {
+    case NO_PROVIDER:
+      console.log('No provider available.');
+      break;
+    case CONNECTION_DENIED:
+      console.log('The user rejected the request to connect with your dApp');
+      break;
+  }
+});
+</pre>
+
+
+'''Provider Request Handling'''
+
+Upon receiving this request, return:
+- The address and public key of account currently selected by the user in the wallet interface
+
+Example
+<pre>
+{
+  address: 'ATUaTd3LA4kZiyB6it9fdb5oJpZYMBF4DX',
+  publicKey: '03fa41b6ff75ebeff8464556629cfceae7402f5d815626a7a6542f786974b942e0'
+}
+</pre>
+
+
+====getBalance====
+Return balance of a specific asset for the given account.
+
+If the asset is omitted from a request to MainNet, all asset and token balances will be returned.
+
+
+'''Method Interface'''
+
+<pre>
+function getBalance(args: GetBalanceArgs): Promise<BalanceResults>
+</pre>
+
+
+'''Input arguments'''
+
+<pre>
+interface GetBalanceArgs {
+  params: BalanceRequest|BalanceRequest[];
+  network?: string - Network to submit this request to. If omitted, will default to network the wallet is currently set to.
+}
+
+interface BalanceRequest {
+  address: string; // Address to check balance(s)
+  assets?: string[]; // Asset IDs or script hashes to check balance.
+  fetchUTXO?: boolean; // Fetches to UTXO data for NEO and/or GAS if attribute is 'true'
+}
+</pre>
+
+
+'''Success return value'''
+
+<pre>
+interface BalanceResults {
+  [address: string]: Balance[];
+}
+
+interface Balance {
+  assetID: string;
+  symbol: string;
+  amount: string;
+}
+</pre>
+
+
+'''Error return value'''
+
+<pre>
+interface Error {
+  type: string; // `NO_PROVIDER`|`CONNECTION_DENIED`
+  description: string;
+  data: string;
+}
+</pre>
+
+
+'''Example'''
+
+<pre>
+getBalance({
+  params: {
+    address: 'AeysVbKWiLSuSDhg7DTzUdDyYYKfgjojru',
+    assets: ['c36aee199dbba6c3f439983657558cfb67629599']
+  },
+})
+.then((results: BalanceResults) => {
+  Object.keys(results).forEach(address => {
+    const balances = results[address];
+    balances.forEach(balance => {
+      const { assetID, symbol, amount } = balance
+
+      console.log('Address: ' + address);
+      console.log('Asset ID: ' + assetID);
+      console.log('Asset symbol: ' + symbol);
+      console.log('Amount: ' + amount);
+    });
+  });
+})
+.catch(({type: string, description: string, data: any}) => {
+  switch(type) {
+    case NO_PROVIDER:
+      console.log('No provider available.');
+      break;
+    case CONNECTION_DENIED:
+      console.log('The user rejected the request to connect with your dApp');
+      break;
+  }
+});
+</pre>
+
+Single asset balance request sample
+<pre>
+// input
+{
+  params: {
+    address: 'AeysVbKWiLSuSDhg7DTzUdDyYYKfgjojru',
+    assets: ['c36aee199dbba6c3f439983657558cfb67629599']
+  },
+}
+
+// output
+{
+  'AeysVbKWiLSuSDhg7DTzUdDyYYKfgjojru': [
+    {
+      'assetID': 'c36aee199dbba6c3f439983657558cfb67629599',
+      'symbol': 'NKN',
+      'amount': '0.00000233',
```

Comment:
This is an example... it should be `233` on amount, and `digits=8`. 
I see it's really string here, so it would work, but I still prefer a numeric response here, to avoid mistakes on parsing.

Reply:
This is left as a string precisely to circumvent the shortcomings of JS float calculations.

Reply:
Ok,may close this one.